File: thz-lang-engine/src/repl.ts
import * as readline from 'readline';
import { ThzLexer } from './lexer.js';
import { ThzParser } from './parser.js';
import { AnalisadorSemantico } from './analisador.js';
import { InterpretadorThz } from './interpretador.js';
import { formatarErroComCaret } from './errors.js';

/* ============================================================
 * THZ REPL v2.2 — sessão persistente com reconstrução do programa
 *
 * Digite comandos THZ (VARIAVEL, EXIBA, SE..., ESTRUTURA, ENUMERACAO).
 * Termine o bloco com uma linha vazia para avaliar.
 *
 *   .ajuda   mostra os comandos
 *   .limpar  reinicia a sessão
 *   .codigo  exibe o programa-fonte acumulado
 *   .sair    encerra
 * ============================================================ */

interface EstadoSessao {
  declaracoesTopo: string[];
  corpo: string[];
}

const CABECALHO_LINHAS = 3; // pragma + PROGRAMA + (vazio)

function montarPrograma(sessao: EstadoSessao): { fonte: string; linhaInicioCorpo: number } {
  const partes: string[] = ['VERSAO_LINGUAGEM "2.2"', 'PROGRAMA SESSAO'];
  for (const decl of sessao.declaracoesTopo) {
    partes.push(decl);
  }
  partes.push('REGRA_NEGOCIO Sessao', 'OPERACAO Principal() : DECIMAL(38, 10)', 'INICIO');
  const linhaInicioCorpo = partes.length + 1; // 1-based: próxima linha após INICIO
  for (const cmd of sessao.corpo) {
    partes.push(cmd);
  }
  partes.push('FIM', 'FIM_REGRA_NEGOCIO', 'FIM_PROGRAMA');
  return { fonte: partes.join('\n'), linhaInicioCorpo };
}

/** Traduz posição do fonte gerado para o trecho digitado pelo usuário. */
function mapearParaChunk(
  linhaGerada: number,
  colunaGerada: number,
  linhaInicioCorpo: number,
  indicePrimeiraLinhaChunk: number,
  chunk: string
): { linha: number; coluna: number } | undefined {
  const linhasCorpoAntes = linhaGerada - linhaInicioCorpo;
  const linhasChunk = chunk.split(/\r?\n/).length;
  if (linhasCorpoAntes < 0 || linhasCorpoAntes >= linhasChunk) return undefined;
  return { linha: linhasCorpoAntes + 1, coluna: colunaGerada };
}

export async function executarRepl(): Promise<void> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    prompt: 'thz> '
  });

  const sessao: EstadoSessao = { declaracoesTopo: [], corpo: [] };
  let buffer: string[] = [];
  let encerrar = false;

  console.log('THZ-LANG REPL v2.2 — digite ".ajuda" para os comandos, ".sair" para encerrar.');
  rl.prompt();

  const avaliarChunk = (chunk: string): void => {
    const indicePrimeiraLinhaChunk = sessao.corpo.length;

    // Declarações de topo viram parte do programa acumulado.
    const inicioTrim = chunk.trimStart();
    if (/^(ESTRUTURA|ENUMERACAO)\b/.test(inicioTrim)) {
      sessao.declaracoesTopo.push(chunk);
      console.log('[SESSAO] Declaração registrada.');
      return;
    }

    sessao.corpo.push(chunk);

    const { fonte, linhaInicioCorpo } = montarPrograma(sessao);
    try {
      const tokens = new ThzLexer(fonte).tokenize();
      const ast = new ThzParser(tokens).parse();

      const erros = new AnalisadorSemantico(ast).analisar({});
      if (erros.length > 0) {
        for (const erro of erros) {
          const posicao = mapearParaChunk(erro.linha, erro.coluna, linhaInicioCorpo, indicePrimeiraLinhaChunk, chunk)
            ?? { linha: erro.linha, coluna: erro.coluna };
          console.error(formatarErroComCaret(chunk, { linha: posicao.linha, coluna: posicao.coluna, mensagem: erro.mensagem }));
        }
        sessao.corpo.pop(); // não polui a sessão com código reprovado
        return;
      }

      const linhas: string[] = [];
      const interpretador = new InterpretadorThz(ast, { saida: (l) => linhas.push(l) });
      interpretador.executarOperacao('Principal', {});
      for (const linha of linhas) console.log(linha);
    } catch (err) {
      const mensagem = (err as Error).message;
      const casamento = /\[Linha (\d+):(\d+)\]/.exec(mensagem);
      if (casamento) {
        const posicao = mapearParaChunk(Number(casamento[1]), Number(casamento[2]), linhaInicioCorpo, indicePrimeiraLinhaChunk, chunk);
        if (posicao) {
          console.error(formatarErroComCaret(chunk, { linha: posicao.linha, coluna: posicao.coluna, mensagem }));
        } else {
          console.error(mensagem);
        }
      } else {
        console.error(mensagem);
      }
      sessao.corpo.pop();
    }
  };

  const processarLinha = (entrada: string): void => {
    const linha = entrada.trim();

    if (buffer.length === 0 && linha.startsWith('.')) {
      switch (linha) {
        case '.sair':
        case '.quit':
          encerrar = true;
          rl.close();
          return;
        case '.limpar':
          sessao.declaracoesTopo.length = 0;
          sessao.corpo.length = 0;
          console.log('[SESSAO] Sessão reiniciada.');
          return;
        case '.codigo':
          console.log(montarPrograma(sessao).fonte || '(vazio)');
          return;
        case '.ajuda':
          console.log([
            'Bloco de comandos + <enter> em linha vazia avalia o bloco.',
            '.ajuda  esta ajuda',
            '.limpar reinicia a sessão',
            '.codigo exibe o programa acumulado',
            '.sair   encerra o REPL'
          ].join('\n'));
          return;
        default:
          console.log("[REPL] Comando desconhecido: '" + linha + "'. Use .ajuda.");
          return;
      }
    }

    if (linha === '') {
      if (buffer.length > 0) {
        avaliarChunk(buffer.join('\n'));
        buffer = [];
      }
      return;
    }

    buffer.push(entrada);
  };

  rl.on('line', (entrada) => {
    processarLinha(entrada);
    if (!encerrar) rl.prompt();
  });

  rl.on('close', () => {
    console.log('\n[SESSAO] Encerrada. Arena liberada.');
  });
}

File: thz-lang-engine-JVM/src/main/java/thz/lang/repl/Repl.java
package thz.lang.repl;

import thz.lang.diagnosticos.DiagnosticoEntrada;
import thz.lang.diagnosticos.Diagnosticos;
import thz.lang.interpretador.InterpretadorThz;
import thz.lang.lexico.ThzLexer;
import thz.lang.semantico.AnalisadorSemantico;
import thz.lang.semantico.ErroSemantico;
import thz.lang.sintatico.ThzParser;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public final class Repl {
    private Repl(){}
    public static void executar() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
        List<String> declaracoesTopo = new ArrayList<>();
        List<String> corpo = new ArrayList<>();
        List<String> buffer = new ArrayList<>();
        System.out.println("THZ-LANG REPL v2.3 — digite \".ajuda\" para os comandos, \".sair\" para encerrar.");
        System.out.print("thz> ");
        System.out.flush();
        String linha;
        while ((linha = br.readLine()) != null) {
            String trim = linha.trim();
            if (buffer.isEmpty() && trim.startsWith(".")) {
                switch (trim) {
                    case ".sair", ".quit" -> { System.out.println("\n[SESSAO] Encerrada. Arena liberada."); return; }
                    case ".limpar" -> { declaracoesTopo.clear(); corpo.clear(); System.out.println("[SESSAO] Sessão reiniciada."); }
                    case ".codigo" -> System.out.println(montarPrograma(declaracoesTopo, corpo));
                    case ".ajuda" -> System.out.println("Bloco de comandos + <enter> em linha vazia avalia o bloco.\n.ajuda  esta ajuda\n.limpar reinicia a sessão\n.codigo exibe o programa acumulado\n.sair   encerra o REPL");
                    default -> System.out.println("[REPL] Comando desconhecido: '" + trim + "'. Use .ajuda.");
                }
                System.out.print("thz> "); System.out.flush(); continue;
            }
            if (trim.isEmpty()) {
                if (!buffer.isEmpty()) {
                    String chunk = String.join("\n", buffer);
                    buffer.clear();
                    avaliarChunk(chunk, declaracoesTopo, corpo);
                }
                System.out.print("thz> "); System.out.flush(); continue;
            }
            buffer.add(linha);
        }
        System.out.println("\n[SESSAO] Encerrada. Arena liberada.");
    }

    private static String montarPrograma(List<String> topo, List<String> corpo) {
        List<String> partes = new ArrayList<>();
        partes.add("VERSAO_LINGUAGEM \"2.3\"");
        partes.add("PROGRAMA SESSAO");
        partes.addAll(topo);
        partes.add("REGRA_NEGOCIO Sessao");
        partes.add("OPERACAO Principal() : DECIMAL(38, 10)");
        partes.add("INICIO");
        partes.addAll(corpo);
        partes.add("FIM");
        partes.add("FIM_REGRA_NEGOCIO");
        partes.add("FIM_PROGRAMA");
        return String.join("\n", partes);
    }

    private static void avaliarChunk(String chunk, List<String> topo, List<String> corpo) {
        String t = chunk.trim();
        if (t.startsWith("ESTRUTURA") || t.startsWith("ENUMERACAO")) { topo.add(chunk); System.out.println("[SESSAO] Declaração registrada."); return; }
        corpo.add(chunk);
        String fonte = montarPrograma(topo, corpo);
        int linhaInicioCorpo = 3 + topo.size() + 3 + 1;
        try {
            var tokens = new ThzLexer(fonte).tokenize();
            var ast = new ThzParser(tokens).parse();
            var erros = new AnalisadorSemantico(ast).analisar();
            if (!erros.isEmpty()) {
                for (ErroSemantico e : erros) System.err.println(Diagnosticos.formatarErroComCaret(chunk, new DiagnosticoEntrada(e.linha() - linhaInicioCorpo + 1, e.coluna(), e.mensagem())));
                corpo.remove(corpo.size()-1);
                return;
            }
            List<String> saidas = new ArrayList<>();
            var interp = new InterpretadorThz(ast, saidas::add, null);
            interp.executarOperacao("Principal", java.util.Map.of());
            for (String s : saidas) System.out.println(s);
        } catch (Exception ex) {
            String msg = ex.getMessage()!=null?ex.getMessage():"";
            var m = java.util.regex.Pattern.compile("\\[Linha (\\d+):(\\d+)\\]").matcher(msg);
            if (m.find()) {
                int gl = Integer.parseInt(m.group(1)), gc=Integer.parseInt(m.group(2));
                int local = gl - linhaInicioCorpo + 1;
                if (local>=1 && local<=chunk.split("\n",-1).length) System.err.println(Diagnosticos.formatarErroComCaret(chunk, new DiagnosticoEntrada(local, gc, msg)));
                else System.err.println(msg);
            } else System.err.println(msg);
            corpo.remove(corpo.size()-1);
        }
    }
}
